### PR TITLE
Salted Password Hashing & Hash Verification Functions Added

### LINKED ISSUE #150: Hash passwords more securely using salted password hashing (pbkdf_hmac along with the SHA256)
# Expected Behavior

Passwords are more secure when hashed with salt and tend to be much harder to crack. Salted Password Hashing with SHA256 & pbkdf_hmac2 is very secure and widely used.

# Current Behavior

Currently the admin site hashes passwords only using the SHA256. Adding another level of complexity with salt and the pbkdf_hmac2 will improve security

```
HASHING: 
def hash_pwd(pwd : str):
    salt = hashlib.sha256(os.urandom(60)).hexdigest().encode('ascii')
    pwd_hash = hashlib.pbkdf2_hmac('sha512', pwd.encode('utf-8'), salt, 100000)
    pwd_hash = binascii.hexlify(pwd_hash)
    final_hashed_pwd = (salt + pwd_hash).decode('ascii')
    return final_hashed_pwd
```

### PR DESCRIPTION
**Pull request by NVombat**
# Description

Added 2 functions and necessary imports:
1) hash_password(password) -> Takes a password, adds salt and hashes the password using SHA512 & PBKDF_HMAC2 algorithms
2) check_hash(email, password) -> Verifies if the password entered by the user matches the hashed password stored in the database

Fixes #150 

## Type of change

- [x] New feature (non-breaking change which adds functionality)

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

Verified it on a test database with user accounts.

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented on my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Review by riju561 — CHANGES_REQUESTED** [review]
I suggest you rebase from the main branch so there are no merge conflicts and call the functions wherever the passwords are being hashed. The passwords are being hashed in 3 files, administrator/models.py, maintainer/models.py and maintainer/views.py.

File: server/githubsrm/administrator/models.py [hunk 1]
```diff
@@ -22,6 +22,11 @@
     ProjectNotFoundError,
 )
 
+
+import hashlib, binascii, os
+
+service = BotoService()
```
Comment (riju561) [inline server/githubsrm/administrator/models.py:28]: No need of line 28

File: server/githubsrm/administrator/models.py [hunk 2]
```diff
@@ -48,6 +53,24 @@ def check_webHook(self, token: Tuple):
             return True
         raise InvalidWebhookError("Invalid token")
 
+
+    def hash_password(self, pwd: str) -> str:
```
Comment (riju561) [inline server/githubsrm/administrator/models.py:57]: remove comments

File: server/githubsrm/administrator/models.py [hunk 3]
```diff
@@ -75,12 +102,33 @@ def verify_admin(self, email: str, password: str) -> bool:
...
             bool
         """
-        hash_password = sha256(password.encode()).hexdigest()
         if value := self.db.admins.find_one(
-            {"$and": [{"password": hash_password}, {"email": email}]}
+                {"email": email}
         ):
-            return value
-        raise InvalidAdminCredentialsError("Invalid credentials")
+            dbpwd = value['password']
```
Comment (riju561) [inline server/githubsrm/administrator/models.py:108]: remove comments

File: server/githubsrm/administrator/views.py [hunk 4]
```diff
@@ -18,14 +18,16 @@
...
-    beta_maintainer_support,
-    get_token,
-    project_pagination,
-    project_single_project,
-    update_token,
+    accepted_project_pagination, alpha_maintainer_support,
+    beta_maintainer_support, get_token, project_pagination,
+    project_single_project
 )
+from .models import (
```
Comment (riju561) [inline server/githubsrm/administrator/views.py:25]: No need of line 25-28

File: server/githubsrm/maintainer/models.py [hunk 5]
```diff
@@ -12,14 +12,78 @@
 from django.conf import settings
 from administrator import jwt_keys
 
+
+import hashlib, binascii, os
+
+from apis.aws import BotoService
```
Comment (riju561) [inline server/githubsrm/maintainer/models.py:18]: No need of line 18 and 20

File: server/githubsrm/maintainer/models.py [hunk 6]
```diff
@@ -12,14 +12,78 @@
...
 from core import service
 
 
+
 class Entry:
     def __init__(self) -> None:
         client = pymongo.MongoClient(settings.DATABASE["mongo_uri"])
         self.db = client[settings.DATABASE["db"]]
 
+    def hash_password(self, pwd: str) -> str:
```
Comment (riju561) [inline server/githubsrm/maintainer/models.py:31]: remove comments

File: server/githubsrm/maintainer/models.py [hunk 7]
```diff
@@ -167,17 +231,18 @@ def set_password(self, key: str, password: str) -> bool:
...
         if "email" not in decode:
             raise AuthenticationErrors("Email not found invalid JWT")
-
+            
         maintainer = self.db.maintainer_credentials.find_one_and_update(
             {"$and": [{"email": decode.get("email")}, {"reset": True}]},
             update={
                 "$set": {
-                    "password": sha256(password.encode()).hexdigest(),
+                    "password": hash_password(password),
```
Comment (riju561) [inline server/githubsrm/maintainer/models.py:245]: Change to self.hash_password(password)

File: server/githubsrm/maintainer/views.py [hunk 8]
```diff
@@ -13,8 +13,17 @@
 
 from . import entry
 from .definitions import MaintainerSchema, RejectionSchema
+from .utils import (
+    RequestSetPassword, project_pagination,
+    project_single_project
+)
+from .models import (
```
Comment (riju561) [inline server/githubsrm/maintainer/views.py:20]: No need of lines 20-23

File: server/githubsrm/maintainer/views.py [hunk 9]
```diff
@@ -166,7 +175,7 @@ def post(self, request, **kwargs) -> JsonResponse:
         if "error" in validate:
             return JsonResponse(data={"error": validate.get("error")}, status=400)
 
-        password_hashed = sha256(request.data["password"].encode()).hexdigest()
+        password_hashed = hash_password(request.data["password"])
```
Comment (riju561) [inline server/githubsrm/maintainer/views.py:178]: Change to entry.hash_password(request.data["password"])

**Review by riju561 — CHANGES_REQUESTED** [review]
Rebase from the main branch and make the required changes

File: server/githubsrm/administrator/models.py [hunk 10]
```diff
@@ -58,13 +75,17 @@ def insert_admin(self, doc: Dict[str, str]) -> bool:
             bool
 
         """
-        if self.db.admins.find_one({"email": doc.get("email")}):
-            raise ExistingAdminError("Admin exists")
-        password = doc.pop("password")
-        hash_password = sha256(password.encode()).hexdigest()
+        if self.db.admins.find_one({"email": doc.get('email')}):
```
Comment (riju561) [inline server/githubsrm/administrator/models.py:78]: We are using custom errors here. Try to change the function to fit the errors. You should find it in the main branch.

File: server/githubsrm/administrator/models.py [hunk 11]
```diff
@@ -75,12 +96,21 @@ def verify_admin(self, email: str, password: str) -> bool:
         Returns:
             bool
         """
-        hash_password = sha256(password.encode()).hexdigest()
         if value := self.db.admins.find_one(
-            {"$and": [{"password": hash_password}, {"email": email}]}
+                {"email": email}
```
Comment (riju561) [inline server/githubsrm/administrator/models.py:100]: We are using custom errors here. Try to change the function to fit the errors. You should find it in the main branch.

**Review by riju561 — CHANGES_REQUESTED** [review]
This should be the last set of changes required.

File: server/githubsrm/administrator/models.py (hunk 11 above)
Comment (NVombat) [reply server/githubsrm/administrator/models.py:100]: What do you mean by changing the function?? Im a little confused

Comment (riju561) [reply server/githubsrm/administrator/models.py:100]: image
Your logic for the function is correct. Just incoporate the custom errors. This was the function previously for your reference. Hope this answers your question.

Comment (NVombat) [reply server/githubsrm/administrator/models.py:100]: Alright thanks just check out what i've done :) Thanks again

**Review by riju561 — APPROVED** [review]

File: server/githubsrm/administrator/models.py [hunk 12]
```diff
@@ -75,12 +96,21 @@ def verify_admin(self, email: str, password: str) -> bool:
...
         ):
```
Comment (Aradhya-Tripathi) [inline server/githubsrm/administrator/models.py:101]: @NVombat please use appropriate linter black, and get rid of all whitespaces

**Review by Aradhya-Tripathi — APPROVED** [review]
LGTM 🚀 ⚡

**Review by riju561 — APPROVED** [review]

**Pull request merged**